### PR TITLE
exposures-init-geometry

### PR DESCRIPTION
**Pull request by emanuel-schmid**
Changes proposed in this PR:
- create geometry and crs within Exposures.__init__
- get rid of raster data (`exp.meta`) 
- introduce properties for cover, deductible, region_id, category_id, geometry, value, latitude, longitude
- introduce methods to get the values (as `np.array`) of the impact function and the assigned centroids by hazard type

This PR fixes #

### PR Author Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] Correct target branch selected (if unsure, select `develop`)
- [ ] Descriptive pull request title added
- [ ] Source branch up-to-date with target branch
- [ ] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [ ] [Tests][testing] updated
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]
- [ ] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Reviewer_Checklist.html) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html
[linter]: https://climada-python.readthedocs.io/en/stable/guide/Guide_Continuous_Integration_and_Testing.html#3.C.--Static-Code-Analysis

### CODE REVIEW

File: climada/entity/exposures/base.py [hunk 1]
```diff
@@ -135,22 +132,183 @@ class Exposures():
...
+        Parameters
+        ----------
+        data : dict, iterable, DataFrame, ndarray
+            data of the initial DataFrame, see ``pandasDataFrame()``
+        index : Index, array
+            index of the initial DataFrame, see ``pandasDataFrame()``
+        columns : Index, array
+            columns of the initial DataFrame, see ``pandasDataFrame()``
+        dtype : dtype
+            data type of the initial DataFrame, see ``pandasDataFrame()``
```
Comment (spjuhel) [inline climada/entity/exposures/base.py:305]:
```suggestion
        data : dict, iterable, DataFrame, ndarray
            data of the initial DataFrame, see ``pandas.DataFrame()``
        index : Index, array
            index of the initial DataFrame, see ``pandas.DataFrame()``
        columns : Index, array
            columns of the initial DataFrame, see ``pandas.DataFrame()``
        dtype : dtype
            data type of the initial DataFrame, see ``pandas.DataFrame()``
```

File: climada/entity/exposures/test/test_base.py [hunk 2]
```diff
@@ -62,36 +63,36 @@ def test_assign_pass(self):
...
+            self.assertEqual(exp.gdf.shape[0], len(exp.hazard_centroids('FL')))
+            np.testing.assert_array_equal(exp.hazard_centroids('FL'), expected_result)
             exp.assign_centroids(Hazard(), overwrite=False)
-            self.assertEqual(exp.gdf.shape[0], len(exp.gdf[INDICATOR_CENTR + 'FL']))
-            np.testing.assert_array_equal(exp.gdf[INDICATOR_CENTR + 'FL'].values, expected_result)
+            self.assertEqual(exp.gdf.shape[0], len(exp.hazard_centroids('FL')))
+            np.testing.assert_array_equal(exp.hazard_centroids('FL'), expected_result)
 
     def test__init__meta_type(self):
         """ Check if meta of type list raises a ValueError in __init__"""
```
Comment (spjuhel) [inline climada/entity/exposures/test/test_base.py:87]:
```suggestion
        """ Check if meta of type list raises a TypeError in __init__"""
```

**Comment by spjuhel** [comment]
Great job ! 

A few comments in addition to the above suggestions:

- I think it would be important to give a more detailed description of what this PR does here and in the changelog. From what I understand, there are more changes than initializing the geometry within the __init__ of Exposure.
- A test is failing, not sure why.

Some conceptual questions:
- The different attributes are now accessed as properties (for instance region_id). This looks cleaner, but doesn't it limit the possibility of adding custom attributes? (Technically, no, because I suppose you can just add new columns to data, but then access become inconsistent). Maybe this isn't a problem.
- Similarly, setting `region_id` or other attributes is not possible after instantiation (or only via exp.data["region_id"] = new_value not consistent), is this wanted?

**Review by spjuhel — COMMENTED** [review]
Great job ! 

A few comments in addition to the suggestion:

- I think it would be important to give a more detailed description of what this PR does here and in the changelog. From what I understand, there are more changes than initializing the geometry within the __init__ of Exposure.
- A test is failing, not sure why.

I have a set of conceptual question:
- The different attributes are now accessed as properties (for instance region_id). This looks cleaner, but doesn't it limit the possibility of adding custom attributes? (Technically, no, because I suppose you can just add new columns to data, but then access become inconsistent). Maybe this isn't a problem.
- Similarly, setting `region_id` or other attributes is not possible after instantiation (or only via exp.data["region_id"] = new_value not consistent)

**Comment by spjuhel** [comment]
I realized you can actually change/add columns directly with `.gdf["colname"]` contrary to what I thought, so ignore this part of my comment. It is actually quite nice that you can't change the GeoDataFrame object as a whole, but that you can change its elements!

I still have a concern if a user tries to change `region_id` or any other property and faces a cold `property 'region_id' of 'Hazard' object has no setter`

Maybe a placeholder setter explaining that you can actually change it through `exp.gdf["region_id"] = whatever` (or another method) would be a nice way to solve this?

**Review by luseverin — COMMENTED** [review]
I have a few comments but those concern mostly updating the docstrings in `exposures/base.py`. In particular, I think it would be good to clarify somewhere what arguments and parameters of the `exposures()` class are now mandatory, and which arguments/parameters are now optional or deprecated. Otherwise, everyhing looks in order, good job!

File: climada/entity/exposures/base.py [hunk 3]
```diff
@@ -102,11 +104,6 @@ class Exposures():
     geometry : pd.Series, optional
         geometry of type Point of each instance.
         Computed in method set_geometry_points().
```
[the remark this answers is not in the document]
Comment (emanuel-schmid) [reply climada/entity/exposures/base.py:106]: well observed! thanks for pointing this out.

File: climada/entity/exposures/base.py [hunk 4]
```diff
@@ -102,11 +104,6 @@ class Exposures():
     geometry : pd.Series, optional
```
[the remark this answers is not in the document]
Comment (emanuel-schmid) [reply climada/entity/exposures/base.py:104]: neither of them is really an attribute but properties. Besides it's not clear what an "optional" attribute could be. I removed them from the pydoc attributes section, together with other properties and methods

File: climada/entity/exposures/base.py [hunk 5]
```diff
@@ -135,22 +132,187 @@ class Exposures():
...
+    @property
+    def _meta(self):
+        """Metadata dictionary, containing raster information derived from geometry"""
+        if not self.data.size:
+            return None
+        _r, meta = u_coord.points_to_raster(self.data)
+        return meta
 
+    @staticmethod
+    def _consolidate(alternative_data, name, value, default=None, equals=lambda x, y: x == y):
```
[the remark this answers is not in the document]
Comment (emanuel-schmid) [reply climada/entity/exposures/base.py:240]: Yes sure.
I was afraid that by adding a pydoc string the method will show up in the official Documentation. Which it shouldn't. It's not meant to be used elsewhere.
But apparently it doesn't, so I add one.

File: climada/entity/exposures/base.py [hunk 6]
```diff
@@ -135,22 +132,183 @@ class Exposures():
...
+            return self.data[INDICATOR_CENTR + haz_type].values
+        if INDICATOR_CENTR in self.data.columns:
+            return self.data[INDICATOR_CENTR].values
+        raise ValueError(f"Missing hazard centroids.")
+
+    @property
+    def _meta(self):
+        """Metadata dictionary, containing raster information derived from geometry"""
+        _r, meta = u_coord.points_to_raster(self.data)
+        return meta
```
[the remark this answers is not in the document]
Comment (emanuel-schmid) [reply climada/entity/exposures/base.py]: I'd rather not. As I understand, there is no intention to make Exposures hold - and thus be responsible for - any kind of raster data. The method is not even called anywhere, I kept it for rather nostalgic reasons, to be reminded of how one extracts raster data from exposures.

**Comment by emanuel-schmid** [comment]
> Maybe a placeholder setter explaining that you can actually change it through `exp.gdf["region_id"] = whatever` (or another method) would be a nice way to solve this?

Actually I'd prefer users to `exp.data["region_id"] = whatever`, `exp.gdf` has changed the name and the property is kept for (better) backwards compatibility. 😉 

But you're right, setter methods for 'region_id', 'category_id' and the other properties are something we should consider. I'd like to delegate this to another PR though.

File: climada/entity/exposures/base.py [hunk 7]
```diff
@@ -135,22 +132,187 @@ class Exposures():
...
+            data of the initial DataFrame, see ``pandas.DataFrame()``
+        index : Index, array
+            index of the initial DataFrame, see ``pandas.DataFrame()``
+        columns : Index, array
+            columns of the initial DataFrame, see ``pandas.DataFrame()``
+        dtype : dtype
+            data type of the initial DataFrame, see ``pandas.DataFrame()``
+        copy : bool
+            Whether to make a copy of the input data, see ``pandas.DataFrame()``
+        geometry : array
```
[the remark this answers is not in the document]
Comment (emanuel-schmid) [reply climada/entity/exposures/base.py]: I've tried to make it clearer in the Parameters section of `__init__` pydoc

File: climada/entity/exposures/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (spjuhel) [reply climada/entity/exposures/base.py:238]: I think I remember that _prefixed functions and methods are hidden by sphinx by default.

File: doc/tutorial/climada_entity_Exposures.ipynb [hunk 8]
```diff
@@ -28,46 +28,57 @@
...
+    "| `geometry` | Point | the geometry column of the `GeoDataFrame`, i.e., latitude (y) and longitude (x) | centroids assignment |\n",
+    "| `value` | float | a value for each exposure | impact calculation |\n",
+    "| `impf_*`  | int | impact functions ids for hazard types.<br>important attribute, since it relates the exposures to the hazard by specifying the impf_act functions.<br>Ideally it should be set to the specific hazard (e.g. `impf_TC`) so that different hazards can be set<br>in the same Exposures (e.g. `impf_TC` and `impf_FL`). | impact calculation |\n",
+    "| `centr_*` | int | centroids index for hazard type.<br>There might be different hazards defined: centr_TC, centr_FL, ...<br>Computed in method `assign_centroids()` | impact calculation |\n",
+    "| `deductible` | float | deductible value for each exposure. <br>Used for insurance | impact calculation |\n",
+    "| `cover` | float | cover value for each exposure. <br>Used for insurance | impact calculation |\n",
+    "| `region_id` | int | region id (e.g. country ISO code) for each exposure | aggregation |\n",
+    "| `category_id` | int | category id (e.g. building code) for each exposure | aggregation |"
```
Comment (spjuhel) [inline doc/tutorial/climada_entity_Exposures.ipynb:52]: Are all these columns required now? If not, maybe you could detail which one are optional.

**Review by spjuhel — COMMENTED** [review]
Apart from the small thing to clarify about which columns of the DataFrame are required or not, I think this is good :)

Great work!

**Comment by emanuel-schmid** [comment]
many thanks for the review @spjuhel ! 🙌

**Pull request merged**